grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -416,6 +427,58 @@ private static Descriptor findTypeDescriptor(FileDescriptor file, FieldDescripto
         return file.findMessageTypeByName(messageType.getName());
     }
 
+    // to make it more efficient, we calculate whether extract response body one time
+    // if there is no matching toplevel field, we set it to null
+    @Nullable
+    private static String calculateResponseBody(List<FieldDescriptor> topLevelFields, String responseBody) {
+        if (StringUtil.isNullOrEmpty(responseBody)) {
+            return null;
+        }
+        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+            if (fieldDescriptor.getName().equals(responseBody)) {
+                return responseBody;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
+        @Nullable final String responseBody = spec.responseBody;
+        if (responseBody == null) {
+            return null;
+        } else {
+            return httpData -> {
+                final byte[] array = httpData.array();
```

Comment:
We return the original `httpData` when exception happens but we create a new `httpData` when there is no exception. So I manually close the `httpData` when there is no exception happens. I think it would break if you use `try` with resource when exception happens because you use the original data but you already close the resource?